### PR TITLE
Delete By Deletion Cost

### LINKED ISSUE #518: [feature request]CloneSet 增加优先级缩容功能 To add the function of scaling down with priority
**What would you like to be added**:
我希望可以在 CloneSet 上增加根据 pod 标签/节点标签 优先删除的能力，不是立即删除，而是在缩容的时候，优先从带有该标签的 pod 中删除实例 或者 从带有该标签的节点上运行的 pod 中删除。
同时可以设置多组标签，多组标签组成一个有序集合，按照集合的顺序，进行 pod 筛选，删除所需数量的 pod

I propose to add a function of priority deletion based on pod label/node label to CloneSet. Instead of immediate deletion, it is preferred to delete instances from PODs with this label or from PODs running on nodes with this label when scaling down. At the same time, this function supports setting up multiple groups of labels, which form an ordered set. After that, the required number of PODs is deleted by POD filtering according to the order of the set.

**Why is this needed**:
k8s 集群底层的计算资源有多种，有的时候我们可能希望优先缩容某一种计算资源，例如同时使用阿里云的 ECS 与 ECI 作为底层计算资源的情况。
另外还有一中在公有云很常见的场景，在平时集群中的节点会由几个包年包月的云服务器组成。而在某些时刻，服务压力增加，我们需要临时扩容一些按量付费的云服务器，分担这段时间的服务压力。在压力过去后，我们希望可以尽快缩容掉这些 云服务器，以便节约资源费用。因此我们需要优先缩容掉临时启动的节点上的pod，以便该节点进行资源回收。并且我们可能会启动多台云服务器，缩容的时候，也是希望尽可能的先缩容一台，再缩容另一台，而不是多节点同时进行缩容

There are many kinds of computing resources at the bottom of K8S cluster, and sometimes we may want to give priority to a certain computing resource, such as using ECS and ECI of Ali Cloud at the same time as the underlying computing resources.
Another common scenario in the public cloud is that the nodes in a cluster would be consist of several cloud servers with subscription. At some point, the service pressure increases, and we need to temporarily expand some pay-as-you-go cloud servers to share the service pressure during this period.
After the pressure has passed, we hope to scale down these cloud servers as soon as possible in order to save resource costs. Therefore, we need to give priority to scaling down the pod on the node that is temporarily started in order to reclaim resources on that node. And we may start multiple cloud servers. Instead of scaling down multiple nodes with the same priority at the same time，we hope to scale down one by one.

### PR DESCRIPTION
**Pull request by PokerBySwq**
Ⅰ. Describe what this PR does
Add the function for cloneset as deleting pod based on pod-deletion-cost

Ⅱ. Does this pull request fix one issue?
fixes #518

Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
pkg/controller/cloneset/sync/cloneset_sync_utils_test.go:608
TestSortingActivePodsWithDeletionCost

Ⅳ. Describe how to verify it
Deploy multiple pod by cloneset firstly, then set section of controller.kubernetes.io/pod-deletion-cost. In the end, scale down cloneset to verify whether the scaling-down sort of pod meet expection

Ⅴ. Special notes for reviews

### CODE REVIEW

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 1]
```diff
@@ -216,3 +221,129 @@ func isPodReady(coreControl clonesetcore.Control, pod *v1.Pod, minReadySeconds i
...
+	// PodPending: 0, PodUnknown: 10, PodRunning: 20 ， Not ready: 0, ready: 1
+	m := map[v1.PodPhase]int{v1.PodPending: 0, v1.PodUnknown: 10, v1.PodRunning: 20}
+
+	iScore := m[s[i].Status.Phase] + btoi(podutil.IsPodReady(s[i]))
+	jScore := m[s[j].Status.Phase] + btoi(podutil.IsPodReady(s[j]))
+
+	// 3. not receives traffic < receives traffic  (a score of 21 means receiving traffic)
+	// If a pod receives traffic but one is not, the not receives traffic one is smaller
+	if iScore != jScore && (iScore == ReceiveTraffic || jScore == ReceiveTraffic) {
+		return iScore < jScore
```
Comment (furykerry) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:261]: consider separate phase and ready checking

```
	// 2. PodPending < PodUnknown < PodRunning
	if podPhaseToOrdinal[s.Pods[i].Status.Phase] != podPhaseToOrdinal[s.Pods[j].Status.Phase] {
		return podPhaseToOrdinal[s.Pods[i].Status.Phase] < podPhaseToOrdinal[s.Pods[j].Status.Phase]
	}
	// 3. Not ready < ready
	// If only one of the pods is not ready, the not ready one is smaller
	if podutil.IsPodReady(s.Pods[i]) != podutil.IsPodReady(s.Pods[j]) {
		return !podutil.IsPodReady(s.Pods[i])
	}
```

**Review by furykerry — DISMISSED** [review]

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 2]
```diff
@@ -216,3 +221,122 @@ func isPodReady(coreControl clonesetcore.Control, pod *v1.Pod, minReadySeconds i
...
+
+const (
+	// PodDeletionCost can be used to set to an int32 that represent the cost of deleting
+	// a pod compared to other pods belonging to the same ReplicaSet. Pods with lower
+	// deletion cost are preferred to be deleted before pods with higher deletion cost.
+	// Note that this is honored on a best-effort basis, and so it does not offer guarantees on
+	// pod deletion order.
+	// The implicit deletion cost for pods that don't set the annotation is 0, negative values are permitted.
+	//
+	// This annotation is beta-level and is only honored when PodDeletionCost feature is enabled.
```
Comment (FillZpp) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:236]: please remove this line, we need not to define the feature gate in Kruise

File: pkg/controller/cloneset/sync/cloneset_sync_utils.go [hunk 3]
```diff
@@ -26,10 +27,14 @@ import (
...
 	"github.com/openkruise/kruise/pkg/util/lifecycle"
 	"github.com/openkruise/kruise/pkg/util/specifieddelete"
+
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	intstrutil "k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/klog"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
 	"k8s.io/utils/integer"
+	"strconv"
```
Comment (FillZpp) [inline pkg/controller/cloneset/sync/cloneset_sync_utils.go:37]: please sort the import lines using `goimports`, it has caused the golang-lint ci failure

**Review by FillZpp — COMMENTED** [review]
Mostly LG. I notice that your commits may be committed by a incorrect account, please update it:

```
$ git config user.name "Github account Name"
$ git config user.email "Github account email"
$ git rebase -i HEAD~3      # use fixup to squash your 3 commits
$ git commit --amend -s --author="Github account Name <Github account email>"
```

**Review by FillZpp — APPROVED** [review]
LGTM

**Pull request merged**